File: server/Services/WatchersService.cs




namespace postitfullstack.Services;

public class WatchersService
{
    private readonly WatchersRepository _watchersRepository;

    public WatchersService(WatchersRepository watchersRepository)
    {
        _watchersRepository = watchersRepository;
    }

    internal Watcher CreateWatcher(Watcher watcherData)
    {
        Watcher watcher = _watchersRepository.CreateWatcher(watcherData);
        return watcher;
    }

    private Watcher GetWatcherById(int watcherId)
    {
        Watcher watcher = _watchersRepository.GetWatcherById(watcherId);

        if (watcher == null)
        {
            throw new Exception("Invalid watcher id: " + watcherId);
        }
        return watcher;
    }

    internal void DeleteWatcher(int watcherId, Account userInfo)
    {
        Watcher watcher = GetWatcherById(watcherId);

        if (watcher.AccountId != userInfo.Id)
        {
            throw new Exception($"YOU CANNOT DELETE ANOTHER USER'S WATCHER, {userInfo.Name.ToUpper()}!!!");
        }

        _watchersRepository.DeleteWatcher(watcherId);
    }

    internal List<WatcherAlbum> GetMyWatcherAlbum(string accountId)
    {
        List<WatcherAlbum> watcherAlbums = _watchersRepository.GetMyWatcherAlbum(accountId);
        return watcherAlbums;
    }

    internal List<WatcherProfile> GetWatcherProfilesByAlbumId(int albumId)
    {
        List<WatcherProfile> watcherProfiles = _watchersRepository.GetWatcherProfilesByAlbumId(albumId);
        return watcherProfiles;
    }
}

File: client/src/services/WatchersService.js
import { logger } from "@/utils/Logger.js"
import { api } from "./AxiosService.js"
import { AppState } from "@/AppState.js"
import { WatcherAlbum, WatcherProfile } from "@/models/Watcher.js"

class WatchersService {
  async createWatcher(watcherData) {
    const response = await api.post('api/watchers', watcherData)
    logger.log('CREATED WATCHER', response.data)
    const watcher = new WatcherProfile(response.data)
    AppState.watcherProfiles.push(watcher)
    AppState.activeAlbum.watcherCount++
  }
  async getWatchersByAlbumId(albumId) {
    AppState.watcherProfiles = []
    const response = await api.get(`/api/albums/${albumId}/watchers`)
    logger.log('GOT WATCHERS', response.data)
    // NOTE make sure you map into the object that supports the correct version of the many-to-many
    const watcherProfiles = response.data.map(pojo => new WatcherProfile(pojo))
    AppState.watcherProfiles = watcherProfiles
  }
  async getMyWatchedAlbums() {
    AppState.watcherAlbums = []
    const response = await api.get('account/watchers')
    logger.log('GOT MY ALBUMS I AM WATCHING', response.data)
    // NOTE make sure you map into the object that supports the correct version of the many-to-many
    const watcherAlbums = response.data.map(pojo => new WatcherAlbum(pojo))
    AppState.watcherAlbums = watcherAlbums
  }
  async deleteWatcher(watcherId) {
    const response = await api.delete(`/api/watchers/${watcherId}`)
    logger.log('DELETED WATCHER', response.data)
    const watcherAlbums = AppState.watcherAlbums
    const index = watcherAlbums.findIndex(watcher => watcher.id == watcherId)
    watcherAlbums.splice(index, 1)
  }
}

export const watchersService = new WatchersService()
